spec/forms/candidate_interface/science_gcse_grade_form_spec.rb

```diff
@@ -0,0 +1,276 @@
+
+      it 'returns no errors if grade is valid' do
+        valid_grades = ['ABC', 'AB', 'AA', 'abc', 'A B C', 'A-B-C']
+
+        valid_grades.each do |grade|
+          form.grade = grade
+          form.validate
+
+          expect(form.errors[:grade]).to be_empty
+        end
+      end
+
+      it 'return validation error if grade is invalid' do
+        invalid_grades = %w[123 A* XYZ]
+
+        invalid_grades.each do |grade|
+          form.grade = grade
+          form.validate(:grade)
+
+          expect(form.errors[:grade]).to include('Enter a real grade')
+        end
+      end
+    end
+
+    context 'when qualification type is Scottish National 5' do
+      let(:form) { CandidateInterface::ScienceGcseGradeForm.build_from_qualification(qualification) }
+      let(:qualification) { FactoryBot.build_stubbed(:application_qualification, qualification_type: 'scottish_national_5', level: 'gcse') }
+
+      it 'returns no errors if grade is valid' do
+        valid_grades = ['AAA', 'AAB', '765', 'CBD', 'aaa', 'C B D', 'C-B-D']
+
+        valid_grades.each do |grade|
+          form.grade = grade
+          form.validate
+
+          expect(form.errors[:grade]).to be_empty
+        end
+      end
+
+      it 'return validation error if grade is invalid' do
+        invalid_grades = %w[89 AE A*]
+
+        invalid_grades.each do |grade|
+          form.grade = grade
+          form.validate(:grade)
+
+          expect(form.errors[:grade]).to include('Enter a real grade')
+        end
+      end
+    end
+  end
+
+  context 'when saving qualification details' do
+    qualification = ApplicationQualification.new
+    form = CandidateInterface::ScienceGcseGradeForm.build_from_qualification(qualification)
+
+    describe '#save_grades' do
+      it 'return false if not valid' do
+        expect(form.save_grades).to eq(false)
+      end
+
+      it 'updates qualification details if valid' do
+        application_form = create(:application_form)
+        qualification = ApplicationQualification.create(level: 'gcse', application_form: application_form)
+        details_form = CandidateInterface::ScienceGcseGradeForm.build_from_qualification(qualification)
+
+        details_form.grade = 'AB'
+
+        details_form.save_grades
+        qualification.reload
+
+        expect(qualification.grade).to eq('AB')
+      end
+
+      it 'sets grade to other_grade if candidate selected "other"' do
+        application_form = create(:application_form)
+        qualification = ApplicationQualification.create(level: 'gcse', application_form: application_form)
+        details_form = CandidateInterface::ScienceGcseGradeForm.build_from_qualification(qualification)
+
+        details_form.grade = 'other'
+        details_form.other_grade = 'D'
+
+        details_form.save_grades
+        qualification.reload
+
+        expect(qualification.grade).to eq('D')
+      end
+    end
+
+    describe '#save_year' do
```

Comment:
Should we have save_year logic in here? That should still currently be handled by the details from right? Which we should now change to year form (after this piece of work)

Reply:
Yep I'll remove